### PR TITLE
fix: populate metrics from OptimStepResponse

### LINKED ISSUE #275: OptimStepResponse.metrics: should it be populated?
We were hoping to monitor gradient norms during SFT while debugging an issue.  OptimStepResponse.metrics exists in the api and docs, but it's unclear what it should contain. In our runs optim_step().result().metrics seems to always be empty.  Should it be populated? Are grad norms exposed anywhere?

### PR DESCRIPTION
**Pull request by nathon-lee**
fix(supervised): populate metrics from OptimStepResponse

Previously, the result of `optim_step_future` was awaited but ignored in `finish_batch`, causing optimization metrics (like gradient norms) to be dropped.

This commit modifies `tinker_cookbook/supervised/train.py` to:
- Capture the `OptimStepResponse` result from the future.
- Merge its `metrics` field (if present) into the current step's metrics dictionary.

Fixes #275

### CODE REVIEW

**Comment by tyler-griggs** [comment]
Thanks @nathon-lee! I made a slight simplification since the `OptimStepResponse` is a pydantic `BaseModel`, we don't need to check whether it has the `metrics` field. Then I added the same metrics extraction to other training loops that were previously dropping them.
